index/segment/fst/writer.go

```diff
@@ -0,0 +1,361 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fst
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/generated/proto/fstwriter"
+	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/index/segment/fs/encoding"
+	"github.com/m3db/m3ninx/postings/pilosa"
+)
+
+var (
+	defaultInitialPostingsOffsetsMapSize = 1024
+	defaultInitialFSTTermsOffsetsMapSize = 1024
+	defaultInitialIntEncoderSize         = 128
+
+	errUnableToFindPostingsOffset = errors.New("internal error: unable to find postings offset")
+	errUnableToFindFSTTermsOffset = errors.New("internal error: unable to find fst terms offset")
+)
+
+type writer struct {
+	seg segment.Segment
+
+	intEncoder      *encoding.Encoder
+	postingsEncoder *pilosa.Encoder
+	fstWriter       *fstWriter
+
+	metadata            []byte
+	docsDataFileWritten bool
+	postingsFileWritten bool
+	fstTermsFileWritten bool
+	postingsOffsets     *postingsOffsetsMap
+	fstTermsOffsets     *fstTermsOffsetsMap
+}
+
+// NewWriter returns a new writer.
+func NewWriter() Writer {
+	return &writer{
+		intEncoder:      encoding.NewEncoder(defaultInitialIntEncoderSize),
+		postingsEncoder: pilosa.NewEncoder(),
+		fstWriter:       newFSTWriter(),
+		postingsOffsets: newPostingsOffsetsMap(defaultInitialPostingsOffsetsMapSize),
+		fstTermsOffsets: newFSTTermsOffsetsMap(defaultInitialFSTTermsOffsetsMapSize),
+	}
+}
+
+func (w *writer) Clear() {
+	w.seg = nil
+
+	w.intEncoder.Reset()
+	w.postingsEncoder.Reset()
+	w.fstWriter.Clear()
```

Comment:
Any reason why fstWriter is Clear not Reset?

Reply:
It has Reset() too, but that requires an argument (io.Writer). It's Clear() releases any held resources, and Reset() makes it ready to write out data.